junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/ExtensionValuesStore.java

```diff
@@ -110,7 +110,16 @@ Object remove(Namespace namespace, Object key) {
 
 	private StoredValue getStoredValue(Namespace namespace, Object key) {
 		CompositeKey compositeKey = new CompositeKey(namespace, key);
-		return this.storedValues.get(compositeKey);
+
+		/*
+		 * This method can be called from another ExtensionValuesStore instance,
+		 * so we must synchronized on the monitor.
+		 *
+		 * TODO: Consider rewriting with ConcurrentHashMap.
```

Comment:
No, I'm saying that when I was working on the rewrite, I realized that the existing code had a thread-visibility issue and a functional bug.

The thread-visibility issue would be less likely to happen if we used `ConcurrentMap`.

I added a commit to fix the bug.